### PR TITLE
Task/inba 480 conditional delete

### PR DESCRIPTION
**Pull request by Perry5**
#### What does this PR do?
Adds check for certain conditions before deleting a subject
https://paper.dropbox.com/doc/Indaba-Conditional-Limits-ublG6ZqtvItRL1LL924Zd#:uid=093953831492899346329275&h2=Conditional-Delete

#### Related JIRA tickets:
https://jira.amida-tech.com/browse/INBA-480

#### How should this be manually tested?
Create a project. Add subjects (delete a subject while you are creating the project)
Delete projects after project has been created. 
Assign task and then delete project. 
Mark task as complete and try deleting a subject. 
etc etc

### CODE REVIEW

File: backend/app/controllers/uoas.js [hunk 1]
```diff
@@ -157,40 +159,115 @@ module.exports = {
...
+
+                            // Soft delete from the UOA table
+                            yield thunkQuery(
+                                'UPDATE "UnitOfAnalysis"' +
+                                'SET "isDeleted" = (to_timestamp(' + Date.now() +
+                                '/ 1000.0)) WHERE "id" = ' + req.params.id
+                            );
+
+                        } else {
+                            if (task[0].isComplete === true) {
```
Comment (sirctseb) [inline backend/app/controllers/uoas.js]: I think this does not quite satisfy the condition in the paper doc. This should check not that the task is complete, but if any question has been answered for the subject. Also, we should be making the check for every task, not just the first

File: backend/app/controllers/uoas.js [hunk 2]
```diff
@@ -157,40 +159,115 @@ module.exports = {
...
+                                );
+                            }
+                        }
+                    }
                 } else {
-                    throw new HttpError(403, 'Cannot delete UOA of already completed task');
+                    throw new HttpError(403, 'No Project Found');
                 }
             } else {
+                // TODO: Replicate checks for conditions if there is no productID passed in.
```
Comment (sirctseb) [inline backend/app/controllers/uoas.js]: Is the work referred to by this TODO going to be done under a separate ticket?

Comment (Perry5) [reply backend/app/controllers/uoas.js]: Yes

File: backend/app/controllers/uoas.js (hunk 1 above)
Comment (Perry5) [reply backend/app/controllers/uoas.js]: The condition to check whether any question has been answered will be done in a separate ticket. 582.

Do we not care that a task is completed at all? If a task is completed, that means they have answered all questions right?

**Comment by jmfadeley** [comment]
If all other issues are in new tickets, we can probably sign off on this.

**Pull request merged**